pkg/features/feature.go

```diff
@@ -0,0 +1,224 @@
+}
+
+type FeatureProvider struct {
+	Features       v3.FeatureInterface
+	FeaturesLister v3.FeatureLister
+	Fallback       map[string]*bool
+}
+
+// newFeatureGate returns a feature gate with an empty map of features and new k8s FeatureGate
+func newFeatureGate() *featureGate {
+	return &featureGate{
+		feat.NewFeatureGate(),
+		make(map[string]*Feature),
+	}
+}
+
+// Enabled returns whether the Feature is enabled in the global feature gate.
+// This should be primarily used for schema enable function and add feature handler functions
+func (f *Feature) Enabled() bool {
+	feature := feat.Feature(f.name)
+	return globalGate.Enabled(feature)
+}
+
+// Set assigns given bool value to given feature
+func (fp *FeatureProvider) Set(name string, value bool) error {
+	obj, err := fp.Features.Get(name, v1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to get feature %s: %v", name, err)
+	}
+
+	if obj.Value != nil {
+		*obj.Value = value
+	} else {
+		obj.Value = newBool(value)
+	}
+
+	_, err = fp.Features.Update(obj)
+	if err != nil {
+		return fmt.Errorf("unable to update feature %s: %v", name, err)
+	}
+
+	return nil
+}
+
+// SetAll iterates a feature map, creates all new features and updates existing features with proper defaults
+func (fp *FeatureProvider) SetAll(featuresMap map[string]*Feature) error {
+	fallback := map[string]*bool{}
+
+	for _, feature := range featuresMap {
+		obj, err := fp.Features.Get(feature.name, v1.GetOptions{})
+		if errors.IsNotFound(err) {
+			// since the feature is not found, it will be created
+			newFeature := &v3.Feature{
+				ObjectMeta: v1.ObjectMeta{
+					Name: feature.name,
+				},
+				Default: feature.def,
+				Value:   feature.value,
+			}
+
+			fallback[newFeature.Name] = newFeature.Value
+
+			_, err := fp.Features.Create(newFeature)
+			if err != nil {
+				return fmt.Errorf("could not create feature %s while setting features: %v", feature.name, err)
+			}
+		} else if err != nil {
+			return fmt.Errorf("unable to get feature %s while setting features: %v", feature.name, err)
+		} else {
+			update := false
+			if obj.Default != feature.def {
+				obj.Default = feature.def
+				update = true
+			}
+			if obj.Value != feature.value {
+				obj.Value = feature.value
+				update = true
+			}
+			if update {
+				_, err := fp.Features.Update(obj)
+				if err != nil {
+					return fmt.Errorf("unable to update feature %s while setting features: %v", feature.name, err)
+				}
+			}
+		}
+	}
+
+	fp.Fallback = fallback
+
+	return nil
+}
+
+// Enabled returns whether given feature is enabled
+func (fp *FeatureProvider) Enabled(name string) bool {
+	obj, err := fp.FeaturesLister.Get("", name)
+	if err != nil {
+		if backup := fp.Fallback[name]; backup != nil {
+			return *fp.Fallback[name]
+		}
+		return false
+	}
+	if obj.Value == nil {
+		return obj.Default
+	}
+
+	return *obj.Value
+}
+
+// SetProvider assigns the given provider as the feature provider
+func SetProvider(p *FeatureProvider) error {
```

Comment:
Rename this func, accept management, and setup the FeatureProvider here. Unexport things.